source/common/config/delta_subscription_state.cc

```diff
@@ -0,0 +1,194 @@
+  ack.error_detail_.set_code(Grpc::Status::GrpcStatus::Internal);
+  ack.error_detail_.set_message(e.what());
+  disableInitFetchTimeoutTimer();
+  stats_.update_rejected_.inc();
+  ENVOY_LOG(warn, "delta config for {} rejected: {}", type_url_, e.what());
+  callbacks_.onConfigUpdateFailed(&e);
+}
+
+void DeltaSubscriptionState::handleEstablishmentFailure() {
+  disableInitFetchTimeoutTimer();
+  stats_.update_failure_.inc();
+  stats_.update_attempt_.inc();
+  callbacks_.onConfigUpdateFailed(nullptr);
+}
+
+envoy::api::v2::DeltaDiscoveryRequest DeltaSubscriptionState::getNextRequest() {
+  envoy::api::v2::DeltaDiscoveryRequest request;
+  if (first_request_of_new_stream_) {
+    // initial_resource_versions "must be populated for first request in a stream".
+    // Also, since this might be a new server, we must explicitly state *all* of our subscription
+    // interest.
+    for (auto const& resource : resource_versions_) {
+      // Populate initial_resource_versions with the resource versions we currently have.
+      // Resources we are interested in, but are still waiting to get any version of from the
+      // server, do not belong in initial_resource_versions. (But do belong in new subscriptions!)
+      if (!resource.second.waitingForServer()) {
+        (*request.mutable_initial_resource_versions())[resource.first] = resource.second.version();
+      }
+      // As mentioned above, fill resource_names_subscribe with everything, including names we
+      // have yet to receive any resource for.
+      names_added_.insert(resource.first);
+    }
+    names_removed_.clear();
+    first_request_of_new_stream_ = false;
+  }
+  std::copy(names_added_.begin(), names_added_.end(),
+            Protobuf::RepeatedFieldBackInserter(request.mutable_resource_names_subscribe()));
+  std::copy(names_removed_.begin(), names_removed_.end(),
+            Protobuf::RepeatedFieldBackInserter(request.mutable_resource_names_unsubscribe()));
+  names_added_.clear();
+  names_removed_.clear();
+
+  request.set_type_url(type_url_);
+  request.mutable_node()->MergeFrom(local_info_.node());
+  return request;
+}
+
+void DeltaSubscriptionState::disableInitFetchTimeoutTimer() {
+  if (init_fetch_timeout_timer_) {
+    init_fetch_timeout_timer_->disableTimer();
+    init_fetch_timeout_timer_.reset();
+  }
+}
+
+void DeltaSubscriptionState::setResourceVersion(const std::string& resource_name,
+                                                const std::string& resource_version) {
+  resource_versions_[resource_name] = ResourceVersion(resource_version);
+  resource_names_.insert(resource_name);
+}
+
+void DeltaSubscriptionState::setResourceWaitingForServer(const std::string& resource_name) {
+  resource_versions_[resource_name] = ResourceVersion();
+  resource_names_.insert(resource_name);
+}
+
+void DeltaSubscriptionState::lostInterestInResource(const std::string& resource_name) {
```

Comment:
Nit: prefer `setLostInterestInResource`

Reply:
Good catch, with the past tense this name is very misleading, sounding like a const getter.